File: app/src/test/scala/tripaint/control/action/OpenActionTest.scala
package tripaint.control.action

import tripaint.Color
import tripaint.TriPaintModel
import tripaint.control.Actions
import tripaint.coords.{GridCoords, StorageCoords}
import tripaint.image.RegularImage
import tripaint.image.format.SimpleStorageFormat
import tripaint.infrastructure.FileSystem
import tripaint.view.FileOpenSettings

import munit.FunSuite

import java.io.File

class OpenActionTest extends FunSuite {
  test("OpenAction should open an image") {
    val file = new File("image.png")
    val image = RegularImage.fill(8, 8, Color.Yellow)
    image.setColor(5, 6, Color.Cyan)

    val model =
      TriPaintModel.createNull(8, FileSystem.NullArgs(initialImages = Map(file -> image)))

    Actions.openImage(
      model,
      file,
      FileOpenSettings(StorageCoords(0, 0), SimpleStorageFormat),
      GridCoords(3, 4)
    )

    val actualImage = model
      .imageGrid(GridCoords(3, 4))
      .map(_.storage)
      .map(_.toRegularImage(SimpleStorageFormat))
      .orNull

    assertEquals(actualImage, image)
  }

  test("OpenAction should open an image at an offset") {
    val file = new File("image.png")
    val image = RegularImage.fill(8, 8, Color.Yellow)
    image.setColor(5, 6, Color.Cyan)

    val storedImage = RegularImage.ofSize(9, 10)
    val offset = StorageCoords(1, 2)
    storedImage.pasteImage(offset, image)

    val model =
      TriPaintModel.createNull(8, FileSystem.NullArgs(initialImages = Map(file -> storedImage)))

    Actions.openImage(
      model,
      file,
      FileOpenSettings(offset, SimpleStorageFormat),
      GridCoords(3, 4)
    )

    val actualImage = model
      .imageGrid(GridCoords(3, 4))
      .map(_.storage)
      .map(_.toRegularImage(SimpleStorageFormat))
      .orNull

    assertEquals(actualImage, image)
  }
}

File: app/src/test/kotlin/tripaint/app/OpenActionTest.kt
package tripaint.app

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Nested
import tripaint.color.Color
import tripaint.coords.GridCoords
import tripaint.coords.StorageCoords
import tripaint.grid.ImageGrid
import tripaint.image.ImagePool
import tripaint.image.RegularImage
import tripaint.image.format.SimpleStorageFormat
import java.io.File
import kotlin.test.Test

class OpenActionTest {
    @Nested
    inner class OpenAction {
        @Test
        fun `opens an image`() {
            val file = File("image.png")
            val image = RegularImage.fill(8, 8, Color.Yellow)
            image.setColor(5, 6, Color.Cyan)

            val fileSystem = FileSystem.createNull(FileSystem.NullArgs(initialImages = mapOf(Pair(file, image))))
            val imagePool = ImagePool()
            val imageGrid = ImageGrid(8)

            Actions.openImage(
                fileSystem, imagePool, imageGrid,
                file,
                Pair(StorageCoords.from(0, 0), SimpleStorageFormat),
                GridCoords.from(3, 4)
            )

            val cell = imageGrid.apply(GridCoords.from(3, 4))!!
            val actualImage = cell.storage.toRegularImage(SimpleStorageFormat)

            assertEquals(image, actualImage)
        }

        @Test
        fun `opens an image at an offset`() {
            val file = File("image.png")
            val image = RegularImage.fill(8, 8, Color.Yellow)
            image.setColor(5, 6, Color.Cyan)

            val storedImage = RegularImage.ofSize(9, 10)
            val offset = StorageCoords.from(1, 2)
            storedImage.pasteImage(offset, image)

            val fileSystem = FileSystem.createNull(FileSystem.NullArgs(initialImages = mapOf(Pair(file, storedImage))))
            val imagePool = ImagePool()
            val imageGrid = ImageGrid(8)

            Actions.openImage(
                fileSystem, imagePool, imageGrid,
                file,
                Pair(offset, SimpleStorageFormat),
                GridCoords.from(3, 4)
            )

            val cell = imageGrid.apply(GridCoords.from(3, 4))!!
            val actualImage = cell.storage.toRegularImage(SimpleStorageFormat)

            assertEquals(image, actualImage)
        }
    }
}